Add support for a reducer step that aggregates per-user metrics

csgregorian opened the pull request:
The goal of this PR is to let Dynamometer collect and emit per-user metrics, as opposed to the whole-workload measurements that it currently collects using Counters. This is done by first changing the `AuditReplayMapper` to emit key-value pairs in the form of `(username_type, latency)` where `type` is either `READ` or `WRITE`. Then, a reducer step `AuditReplayReducer` is added to sum latencies per key. More generally, this change allows Dynamometer to support arbitrary reducers for stats aggregation along with the existing mappers.

I modified `TestWorkloadGenerator` to run with and without a reducer, and confirm that the file was created. This still needs a test to ensure that the actual contents of the output file are correct. As well, it's possible that I changed some things that didn't need to be changed in the process of getting the tests to pass/output to appear.

Mostly trying to see if CI will pass, but ready for some preliminary reviews.

EDIT: a couple other things that I'm not sure are the right approach and need to be looked at a bit closer:
- Not sure whether extending `WorkloadMapper` to specify `KEYOUT` and `VALUEOUT` as parameters is the right move
- What happens if `AuditReplayMapper` writes to the context without a reducer being specified? Should `job.setMapOutputKeyClass` be set to `LongWritable.class` instead of `NullWritable.class` regardless of whether it's actually outputting anything? It doesn't seem like this caused any errors, but if not, why?
- I simplified `WorkloadReducer` just to what it needs to run, but it should probably have `getDescription` and `getConfigDescriptions` as well.

xkrogen commented on dynamometer-workload/src/main/java/com/linkedin/dynamometer/workloadgenerator/WorkloadDriver.java:
```diff
@@ -64,6 +68,11 @@ public int run(String[] args) throws Exception {
         .isRequired().create(MAPPER_CLASS_NAME);
     options.addOption(mapperClassOption);
+    Option reducerClassOption = OptionBuilder.withArgName("Reducer ClassName").hasArg().withDescription(
+        "Class name of the reducer (optional); must be a Reducer subclass. Reducers supported currently: \n" +
```
"must be a WorkloadReducer sublcass" ?

xkrogen commented on dynamometer-workload/src/main/java/com/linkedin/dynamometer/workloadgenerator/WorkloadDriver.java:
```diff
@@ -151,9 +177,21 @@ public static void main(String[] args) throws Exception {
     return (Class<? extends WorkloadMapper>) mapperClass;
   }
 
+  private Class<? extends WorkloadReducer> getReducerClass(String className) throws ClassNotFoundException {
```
I wonder if it's possible reduce code duplication between `getReducerClass` and `getMapperClass`?

xkrogen commented on dynamometer-workload/src/main/java/com/linkedin/dynamometer/workloadgenerator/WorkloadReducer.java:
```diff
@@ -0,0 +1,18 @@
+package com.linkedin.dynamometer.workloadgenerator;
```
Needs license header

xkrogen commented on dynamometer-workload/src/main/java/com/linkedin/dynamometer/workloadgenerator/audit/AuditReplayReducer.java:
```diff
@@ -0,0 +1,36 @@
+package com.linkedin.dynamometer.workloadgenerator.audit;
```
needs license header

xkrogen commented on dynamometer-workload/src/main/java/com/linkedin/dynamometer/workloadgenerator/audit/AuditReplayThread.java:
```diff
@@ -254,7 +258,16 @@ public FileSystem run() {
       long latency = System.currentTimeMillis() - startTime;
+
+      try {
+        mapperContext.write(new Text(key), new LongWritable(latency));
```
It's been a while since I've written an MR job, but I believe it's more efficient from a GC perspective to instantiate single instances of `Text` and `LongWritable` and then use their `set` methods to change their contents.

xkrogen commented on dynamometer-workload/src/main/java/com/linkedin/dynamometer/workloadgenerator/audit/AuditReplayThread.java:
```diff
@@ -254,7 +258,16 @@ public FileSystem run() {
           break;
       }
+
+      String key = command.getSimpleUgi() + "_" + replayCommand.getType().toString();
```
Using a String concatenation doesn't need like the most efficient or clean route. I think it should be pretty easy to create a composite key like `class UserCommandKey implements Writable` which has two `Text` fields that it delegates to for serialization.

xkrogen commented on dynamometer-workload/src/main/java/com/linkedin/dynamometer/workloadgenerator/audit/AuditTextOutputFormat.java:
```diff
@@ -0,0 +1,26 @@
+/**
+ * Copyright 2017 LinkedIn Corporation. All rights reserved. Licensed under the BSD-2 Clause license.
```
2019

xkrogen commented on dynamometer-workload/src/main/java/com/linkedin/dynamometer/workloadgenerator/audit/AuditTextOutputFormat.java:
```diff
@@ -0,0 +1,26 @@
+import java.io.IOException;
+import java.util.List;
+
+public class AuditTextOutputFormat extends TextOutputFormat {
```
I don't think we need a dedicated output format just for this. Within the `WorkloadDriver`, you should be able to use `FileOutputFormat.setOutputPath()` for this.

xkrogen commented on dynamometer-workload/src/test/java/com/linkedin/dynamometer/workloadgenerator/TestWorkloadGenerator.java:
```diff
@@ -4,10 +4,8 @@
-import com.linkedin.dynamometer.workloadgenerator.audit.AuditLogDirectParser;
-import com.linkedin.dynamometer.workloadgenerator.audit.AuditLogHiveTableParser;
-import com.linkedin.dynamometer.workloadgenerator.audit.AuditReplayMapper;
+import com.linkedin.dynamometer.workloadgenerator.audit.*;
```
Don't use `*` imports for classes.

xkrogen commented on dynamometer-workload/src/test/java/com/linkedin/dynamometer/workloadgenerator/TestWorkloadGenerator.java:
```diff
@@ -63,15 +61,40 @@ public void testAuditWorkloadDirectParser() throws Exception {
+    conf.set(AuditReplayReducer.OUTPUT_PATH_KEY, auditOutputPath);
+    conf.setLong(AuditLogDirectParser.AUDIT_START_TIMESTAMP_KEY, 60*1000);
+    testAuditWorkloadWithReducer(AuditReplayReducer.class);
+    assertTrue(dfs.exists(new Path(auditOutputPath)));
```
The tests look great, good use of DRY. But let's not just assert the file exists, let's examine its contents as well.

csgregorian commented:
Simplified a ton of stuff here.
- Reducers are local to a WorkloadMapper instead of being user configured
- WorkloadMappers handle their own configuration
- AuditReplayThreads no longer write directly to context, instead aggregating all of their command latencies locally. These are then collected during the AuditReplayMapper cleanup, similar to the `drainCounters` behaviour. This obviates the need for a combiner, as the maximum number of inputs to the Reducer is now `unique keys * threads per mapper * mappers`, instead of `commands per thread * threads per mapper * mappers`
- Changed AuditReplayMapper/Reducer key type to a custom composite writable `userCommandKey` (not simplified, but the right way to do it [probably])
- Extended `TestWorkloadGenerator` to test the output file contents

xkrogen commented on dynamometer-workload/src/main/java/com/linkedin/dynamometer/workloadgenerator/CreateFileMapper.java:
```diff
@@ -13,6 +13,7 @@
 import org.apache.hadoop.fs.Path;
 import org.apache.hadoop.io.LongWritable;
 import org.apache.hadoop.io.NullWritable;
+import org.apache.hadoop.mapreduce.Job;
```
unused import

xkrogen commented on dynamometer-workload/src/main/java/com/linkedin/dynamometer/workloadgenerator/audit/AuditReplayReducer.java:
```diff
@@ -0,0 +1,23 @@
+
+  @Override
+  protected void reduce(UserCommandKey key, Iterable<LongWritable> values, Context context) throws IOException, InterruptedException {
+    long sum = 0;
```
If you want: `long sum = StreamSupport.stream(values.spliterator(), false).mapToLong(LongWritable::get).sum()`

xkrogen replied:
Not convinced that this is better 🤔

xkrogen commented on dynamometer-workload/src/main/java/com/linkedin/dynamometer/workloadgenerator/audit/AuditReplayThread.java:
```diff
@@ -99,6 +101,16 @@ void drainCounters(Mapper.Context context) {
     }
   }
 
+  void drainCommandLatencies(Mapper.Context context) throws IOException {
```
We shouldn't catch and swallow `InterruptedException` here. It should bubble up to the `cleanup()`. Also, no need to re-throw the IOE.

xkrogen commented on dynamometer-workload/src/main/java/com/linkedin/dynamometer/workloadgenerator/audit/UserCommandKey.java:
```diff
@@ -0,0 +1,89 @@
+import org.apache.hadoop.io.WritableComparable;
+
+
+public class UserCommandKey implements WritableComparable {
```
Javadoc?

xkrogen commented on dynamometer-workload/src/main/java/com/linkedin/dynamometer/workloadgenerator/audit/AuditReplayReducer.java:
```diff
@@ -0,0 +1,23 @@
+
+import java.io.IOException;
+
+public class AuditReplayReducer extends
```
Javadoc?

xkrogen commented on dynamometer-workload/src/main/java/com/linkedin/dynamometer/workloadgenerator/audit/UserCommandKey.java:
```diff
@@ -0,0 +1,89 @@
+    return user.toString();
+  }
+
+  public void setUser(String user) {
```
Are these setters used for anything? Does the framework require them?

xkrogen replied:
Or the getters, for that matter?

csgregorian replied:
The getters are necessary for the equality check; the setters can probably be omitted.

xkrogen replied:
sgtm

xkrogen commented on dynamometer-workload/src/main/java/com/linkedin/dynamometer/workloadgenerator/audit/UserCommandKey.java:
```diff
@@ -0,0 +1,89 @@
+
+  @Override
+  public int compareTo(@Nonnull Object o) {
+    return Integer.compare(hashCode(), o.hashCode());
```
Why is this being sorted by its hash code? Why not a lexicographical sort?

xkrogen commented on dynamometer-workload/src/test/java/com/linkedin/dynamometer/workloadgenerator/TestWorkloadGenerator.java:
```diff
@@ -105,5 +111,12 @@ private void testAuditWorkload() throws Exception {
+    FSDataInputStream auditOutput = dfs.open(new Path(auditOutputPath + "/part-r-00000"));
+    byte[] buf = new byte[auditOutput.available()];
+    assertTrue(auditOutput.read(buf) > 0);
+    System.out.println(new String(buf));
```
Shouldn't println here. If you want, convert this to a `LOG.debug(...)`.

xkrogen commented on dynamometer-workload/src/test/java/com/linkedin/dynamometer/workloadgenerator/TestWorkloadGenerator.java:
```diff
@@ -105,5 +111,12 @@ private void testAuditWorkload() throws Exception {
     assertFalse(dfs.exists(new Path("/denied")));
+
+    assertTrue(dfs.exists(new Path(auditOutputPath)));
+    FSDataInputStream auditOutput = dfs.open(new Path(auditOutputPath + "/part-r-00000"));
```
Use try-with-resources to make sure this is closed.

csgregorian replied:
TIL, cool

xkrogen commented on dynamometer-workload/src/test/java/com/linkedin/dynamometer/workloadgenerator/TestWorkloadGenerator.java:
```diff
@@ -105,5 +111,12 @@ private void testAuditWorkload() throws Exception {
+
+    assertTrue(dfs.exists(new Path(auditOutputPath)));
+    FSDataInputStream auditOutput = dfs.open(new Path(auditOutputPath + "/part-r-00000"));
+    byte[] buf = new byte[auditOutput.available()];
```
Might be easier to use `IOUtils.toString(auditOutput)`

xkrogen commented on dynamometer-workload/src/test/java/com/linkedin/dynamometer/workloadgenerator/TestWorkloadGenerator.java:
```diff
@@ -105,5 +111,12 @@ private void testAuditWorkload() throws Exception {
+    byte[] buf = new byte[auditOutput.available()];
+    assertTrue(auditOutput.read(buf) > 0);
+    System.out.println(new String(buf));
+    assertTrue(new String(buf).matches(".*hdfs,WRITE\\t[0-9]+\\n.*"));
```
Since we actually know how many commands to expect for every user, I would prefer to assert on exact output. Then we ensure the counting is working properly.

csgregorian replied:
The reducer output is the summed latencies, so it's not exactly consistent across runs.

xkrogen replied:
Oh, right. I was thinking it was a count. Okay, SGTM.

xkrogen commented on dynamometer-workload/src/test/java/com/linkedin/dynamometer/workloadgenerator/TestWorkloadGenerator.java:
```diff
@@ -113,10 +112,9 @@ private void testAuditWorkloadWithOutput(String auditOutputPath) throws Exceptio
-    assertTrue(auditOutput.read(buf) > 0);
-    System.out.println(new String(buf));
-    assertTrue(new String(buf).matches(".*hdfs,WRITE\\t[0-9]+\\n.*"));
+    try (FSDataInputStream auditOutputFile = dfs.open(new Path(auditOutputPath + "/part-r-00000"))) {
```
Minor nit: the correct way to make a child path is `new Path(auditOutputPath, "part-r-0000")`

csgregorian replied:
Ah I noticed this one too and forgot to change it 👍 fixed

xkrogen commented on dynamometer-workload/src/main/java/com/linkedin/dynamometer/workloadgenerator/audit/UserCommandKey.java:
```diff
@@ -12,7 +12,11 @@
 
-
+/**
+ * <p>UserCommandKey is a {@link Writable} used as a composite key combining the user id and
```
No need to start with a `<p>`. Only needed to make a subsequent paragraph break.

xkrogen commented on dynamometer-workload/src/main/java/com/linkedin/dynamometer/workloadgenerator/audit/AuditReplayReducer.java:
```diff
@@ -9,6 +9,12 @@
 
+
+/**
+ * <p>AuditReplayReducer aggregates the returned latency values from {@link AuditReplayMapper} and sums
```
No need to start with a `<p>`. Only needed to make a subsequent paragraph break.
